File: rust/crates/orca-config/src/feature_education_telemetry.rs
//! Feature-education telemetry tables + membership-fallback normalizers, ported
//! from `src/shared/feature-education-telemetry.ts`.
//!
//! These tables bound the dimension values emitted to analytics. The normalizers
//! map any off-table string (or null/undefined) to a safe in-table fallback so a
//! caller can never leak an arbitrary string (e.g. a URL) into telemetry.

use crate::contextual_tours::ContextualTourId;

/// Mirrors `contextual_tours::CONTEXTUAL_TOUR_IDS`; the unit test pins the two
/// lists together so telemetry stays aligned with the tour catalog.
pub const FEATURE_EDUCATION_CONTEXTUAL_TOUR_IDS: [ContextualTourId; 6] = [
    ContextualTourId::WorkspaceBoard,
    ContextualTourId::WorkspaceAgentSessions,
    ContextualTourId::Browser,
    ContextualTourId::Tasks,
    ContextualTourId::Automations,
    ContextualTourId::WorkspaceCreation,
];

pub const FEATURE_EDUCATION_SOURCES: [&str; 9] = [
    "workspace_board_visible",
    "workspace_agent_sessions_visible",
    "browser_visible",
    "tasks_open",
    "automations_open",
    "workspace_creation_visible",
    "workspace_creation_modal",
    "setup_guide_parallel_work",
    "unknown",
];

pub const CONTEXTUAL_TOUR_OUTCOMES: [&str; 3] = ["completed", "skipped", "cancelled"];

pub const SETUP_GUIDE_SOURCES: [&str; 6] =
    ["sidebar", "contextual_tour", "settings", "feature_wall", "help_menu", "unknown"];

pub const SETUP_GUIDE_CLOSE_OUTCOMES: [&str; 3] = ["completed", "dismissed", "interrupted"];

pub const TERMINAL_PANE_SPLIT_SOURCES: [&str; 5] =
    ["contextual_tour", "keyboard", "context_menu", "command", "unknown"];

pub fn normalize_feature_education_source(value: Option<&str>) -> &'static str {
    value
        .and_then(|v| FEATURE_EDUCATION_SOURCES.iter().copied().find(|&source| source == v))
        .unwrap_or("unknown")
}

pub fn normalize_setup_guide_source(value: Option<&str>) -> &'static str {
    value
        .and_then(|v| SETUP_GUIDE_SOURCES.iter().copied().find(|&source| source == v))
        .unwrap_or("unknown")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::contextual_tours::CONTEXTUAL_TOUR_IDS;

    #[test]
    fn keeps_contextual_tour_telemetry_ids_aligned_with_tour_definitions() {
        assert_eq!(FEATURE_EDUCATION_CONTEXTUAL_TOUR_IDS, CONTEXTUAL_TOUR_IDS);
    }

    #[test]
    fn normalizes_unknown_telemetry_sources_to_a_bounded_fallback() {
        assert_eq!(normalize_feature_education_source(Some("tasks_open")), "tasks_open");
        assert_eq!(
            normalize_feature_education_source(Some("workspace_agent_sessions_visible")),
            "workspace_agent_sessions_visible"
        );
        assert_eq!(
            normalize_feature_education_source(Some("setup_guide_parallel_work")),
            "setup_guide_parallel_work"
        );
        assert_eq!(normalize_feature_education_source(Some("https://example.com/private")), "unknown");
        assert_eq!(normalize_feature_education_source(None), "unknown");
    }

    #[test]
    fn normalizes_setup_guide_telemetry_sources_to_a_bounded_fallback() {
        assert_eq!(normalize_setup_guide_source(Some("sidebar")), "sidebar");
        assert_eq!(normalize_setup_guide_source(Some("settings")), "settings");
        assert_eq!(normalize_setup_guide_source(Some("help_menu")), "help_menu");
        assert_eq!(normalize_setup_guide_source(Some("private-source")), "unknown");
    }
}

File: src/shared/feature-education-telemetry.ts
import type { ContextualTourId } from './contextual-tours'

export const FEATURE_EDUCATION_CONTEXTUAL_TOUR_IDS = [
  'workspace-board',
  'workspace-agent-sessions',
  'browser',
  'tasks',
  'automations',
  'floating-workspace',
  'workspace-creation'
] as const satisfies readonly ContextualTourId[]

export const FEATURE_EDUCATION_SOURCES = [
  'workspace_board_visible',
  'workspace_agent_sessions_visible',
  'browser_visible',
  'tasks_open',
  'automations_open',
  'floating_workspace_visible',
  'workspace_creation_visible',
  'workspace_creation_modal',
  'setup_guide_parallel_work',
  'unknown'
] as const

export const CONTEXTUAL_TOUR_OUTCOMES = ['completed', 'skipped', 'cancelled'] as const
export const SETUP_GUIDE_SOURCES = [
  'sidebar',
  'contextual_tour',
  'settings',
  'feature_wall',
  'help_menu',
  'unknown'
] as const
export const SETUP_GUIDE_CLOSE_OUTCOMES = ['completed', 'dismissed', 'interrupted'] as const
export const TERMINAL_PANE_SPLIT_SOURCES = [
  'contextual_tour',
  'keyboard',
  'context_menu',
  'command',
  'unknown'
] as const

export type FeatureEducationSource = (typeof FEATURE_EDUCATION_SOURCES)[number]
export type ContextualTourOutcome = (typeof CONTEXTUAL_TOUR_OUTCOMES)[number]
export type SetupGuideSource = (typeof SETUP_GUIDE_SOURCES)[number]
export type SetupGuideCloseOutcome = (typeof SETUP_GUIDE_CLOSE_OUTCOMES)[number]
export type TerminalPaneSplitSource = (typeof TERMINAL_PANE_SPLIT_SOURCES)[number]

// normalizeFeatureEducationSource / normalizeSetupGuideSource were cut over to
// the Rust orca-config core: the renderer drives them via the orca-git wasm
// wrapper in src/renderer/src/lib/git-wasm/feature-education-telemetry.ts. The
// enum DATA consts + types above stay here (imported as z.enum by main
// terminal.ts and shared telemetry-events.ts) with NO napi/wasm dependency.
